src/audio/kpb.c

```diff
@@ -2378,6 +2399,192 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
+		return 0;
+	}
+	return -EINVAL;
+}
+
+static void devicelist_reset(struct device_list *devlist, bool remove_items)
+{
+	/* deallocate items */
+	if (remove_items) {
+		for (int i = 0; i < DEVICE_LIST_SIZE; i++)
+			*devlist->devs[i] = NULL;
+	}
+	/* zero the pointers */
+	for (int i = 0; i < DEVICE_LIST_SIZE; i++)
+		devlist->devs[i] = NULL;
+
+	devlist->count = 0;
+}
+
+static struct comp_dev *get_dev_from_mi_id(uint32_t module_id, uint32_t instance_id)
+{
+	struct comp_dev *dev = NULL;
+	/* for now considering only IPC4 */
+#ifdef CONFIG_IPC_MAJOR_4
+	uint32_t comp_id = IPC4_COMP_ID(module_id, instance_id);
+
+	dev = ipc4_get_comp_dev(comp_id);
+#endif
+	return dev;
+}
+
+static inline int alloc_fmt_module_list_item(struct kpb_fmt_dev_list *fmt_device_list,
+					     struct comp_dev *mi_ptr, struct comp_dev ***item)
+{
+	int first_empty_slot_idx = -1;
+
+	for (size_t module_slot_idx = 0; module_slot_idx < FAST_MODE_TASK_MAX_MODULES_COUNT;
+	     ++module_slot_idx){
+		/* check if module already added */
+		if (fmt_device_list->modules_list_item[module_slot_idx] == mi_ptr)
+			/* was ADSP_INVALID_REQUEST; */
+			return -EINVAL;
+		/* finding first available empty slot */
+		if (first_empty_slot_idx < 0 &&
+		    !fmt_device_list->modules_list_item[module_slot_idx])
+			first_empty_slot_idx = module_slot_idx;
+	}
+	/* add item first available empty slot */
+	if (first_empty_slot_idx >= 0) {
+		fmt_device_list->modules_list_item[first_empty_slot_idx] = mi_ptr;
+		*item = &fmt_device_list->modules_list_item[first_empty_slot_idx];
+		return 0;
+	}
+	return -ENOMEM;
+}
+
+static int prepare_fmt_modules_list(struct comp_dev *kpb_dev,
+				    uint32_t outpin_idx,
+				    const struct kpb_task_params *modules_to_prepare,
+				    struct comp_dev **last_copier_mi)
+{
+	if (!kpb_dev)
+		return -EINVAL;
+	if (!modules_to_prepare)
+		return -EINVAL;
+	if (!last_copier_mi)
+		return -EINVAL;
+	if (outpin_idx >= KPB_MAX_SINK_CNT)
+		return -EINVAL;
+
+	int ret = 0;
+	struct comp_dev *dev = NULL;
+	struct kpb_fmt_dev_list *fmt_device_list =
+			&((struct comp_data *)comp_get_drvdata(kpb_dev))->fmt_device_list;
+
+	if (modules_to_prepare->number_of_modules != 0) {
+		fmt_device_list->kpb_list_item[outpin_idx] = kpb_dev;
+		ret = devicelist_push(&fmt_device_list->device_list[outpin_idx],
+				      &fmt_device_list->kpb_list_item[outpin_idx]);
+		if (ret < 0)
+			return ret;
+	}
+	for (size_t mod_idx = 0; mod_idx < modules_to_prepare->number_of_modules; ++mod_idx) {
+		dev = get_dev_from_mi_id(modules_to_prepare->dev_ids[mod_idx].module_id,
+					 modules_to_prepare->dev_ids[mod_idx].instance_id);
+
+		if (!dev)
+			return -EINVAL;
+
+		struct comp_dev **new_list_item_ptr;
+
+		ret = alloc_fmt_module_list_item(fmt_device_list, dev, &new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+		*new_list_item_ptr = dev;
+		ret = devicelist_push(&fmt_device_list->device_list[outpin_idx],
+				      new_list_item_ptr);
+		if (ret < 0)
+			return ret;
+	}
```

Comment:
I didn't read that much into the original FMT functionality code yet, but I do have some info about it though. In reference FW, executing the FMT also triggered processing of some LL modules whenever there are available DSP cycles. I'm guessing those modules are the ones provided in the IPC here.

I agree that the way the lists were (and are) handled was kind of confusing, but I'm not sure yet what can I refactor here. It's either some important logic, or just some C++'ism.

The idea for now was to implement it up to this point so that the tests related to the FMT config IPC pass. After FMT functionality actually gets implemented it will be easier to see if we can simplify the config.